lsat/interceptor.go

```diff
@@ -0,0 +1,201 @@
+			macaroons.NewMacaroonCredential(macaroon),
+		))
+		return nil
+	}
+
+	// If we already have a token, let's append it.
+	if i.store.HasToken() {
+		lsat, err := i.store.Token()
+		if err != nil {
+			return err
+		}
+		if err = addLsatCredentials(lsat); err != nil {
+			return err
+		}
+	}
+
+	// We need a way to extract the response headers sent by the
+	// server. This can only be done through the experimental
+	// grpc.Trailer call option.
+	// We execute the request and inspect the error. If it's the
+	// LSAT specific payment required error, we might execute the
+	// same method again later with the paid LSAT token.
+	trailerMetadata := &metadata.MD{}
+	opts = append(opts, grpc.Trailer(trailerMetadata))
+	err := invoker(ctx, method, req, reply, cc, opts...)
+
+	// Only handle the LSAT error message that comes in the form of
+	// a gRPC status error.
+	if isPaymentRequired(err) {
+		lsat, err := i.payLsatToken(ctx, trailerMetadata)
+		if err != nil {
+			return err
+		}
+		if err = addLsatCredentials(lsat); err != nil {
+			return err
+		}
+
+		// Execute the same request again, now with the LSAT
+		// token added as an RPC credential.
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+	return err
+}
+
+// payLsatToken reads the payment challenge from the response metadata and tries
+// to pay the invoice encoded in them, returning a paid LSAT token if
+// successful.
+func (i *Interceptor) payLsatToken(ctx context.Context, md *metadata.MD) (
+	*Token, error) {
+
+	// First parse the authentication header that was stored in the
+	// metadata.
+	authHeader := md.Get(AuthHeader)
+	if len(authHeader) == 0 {
+		return nil, fmt.Errorf("auth header not found in response")
+	}
+	matches := authHeaderRegex.FindStringSubmatch(authHeader[0])
+	if len(matches) != 3 {
+		return nil, fmt.Errorf("invalid auth header "+
+			"format: %s", authHeader[0])
+	}
+
+	// Decode the base64 macaroon and the invoice so we can store the
+	// information in our store later.
+	macBase64, invoiceStr := matches[1], matches[2]
+	macBytes, err := base64.StdEncoding.DecodeString(macBase64)
+	if err != nil {
+		return nil, fmt.Errorf("base64 decode of macaroon failed: "+
+			"%v", err)
+	}
+	invoice, err := zpay32.Decode(invoiceStr, i.lnd.ChainParams)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode invoice: %v", err)
+	}
+
+	// Pay invoice now and wait for the result to arrive or the main context
+	// being canceled.
+	respChan := i.lnd.Client.PayInvoice(
+		ctx, invoiceStr, MaxRoutingFeeSats, nil,
+	)
+	select {
+	case result := <-respChan:
+		if result.Err != nil {
+			return nil, result.Err
+		}
+		token, err := NewToken(
+			macBytes, invoice.PaymentHash, result.Preimage,
+			lnwire.NewMSatFromSatoshis(result.PaidAmt),
+			lnwire.NewMSatFromSatoshis(result.PaidFee),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create token: %v",
+				err)
+		}
+		return token, i.store.StoreToken(token)
```

Comment:
Yes, I agree this could happen.
Unfortunately AFAICT we don't have the `TrackPayment` method exposed in the loops's `lndclient` yet.
And since this is pretty unlikely to happen anyway (the client has to be shut down exactly at the moment when the first request to the server is done), I propose to fix that in a later PR.
I added a TODO comment.

Reply:
Ah, I must've looked in the wrong place... You are right then, we should address any gaps we can.
I've added the `trackPayment` functionality and also a unit test.